File: infrastructure/ts-sdk/src/keycloak.ts
/**
 * Keycloak client with token validation, distributed caching with pub/sub invalidation,
 * KYC level extraction, and admin ops.
 */

interface TokenCache { claims: Record<string, unknown>; expiresAt: number; }

export class KeycloakClient {
  private realmUrl: string;
  private clientId: string;
  private clientSecret: string;
  private adminUrl: string;
  private tokenCache = new Map<string, TokenCache>();
  private readonly cacheTTL = 300_000; // 5 minutes
  private redisClient: any = null; // Optional Redis for distributed invalidation

  constructor(realmUrl: string, clientId: string, clientSecret: string, adminUrl: string) {
    this.realmUrl = realmUrl;
    this.clientId = clientId;
    this.clientSecret = clientSecret;
    this.adminUrl = adminUrl;
  }

  /** Attach Redis client for distributed token invalidation across replicas. */
  setRedisClient(redis: any): void {
    this.redisClient = redis;
    this._subscribeToInvalidations();
  }

  private _subscribeToInvalidations(): void {
    if (!this.redisClient?.client) return;
    try {
      const Redis = require('ioredis');
      const [host, port] = (this.redisClient.addr || 'localhost:6379').split(':');
      const sub = new Redis({ host, port: parseInt(port || '6379') });
      sub.subscribe('__token_invalidation__');
      sub.on('message', (_channel: string, message: string) => {
        try {
          const { token, user_id } = JSON.parse(message);
          if (token) {
            this.tokenCache.delete(token);
          }
          if (user_id) {
            // Invalidate all tokens for this user
            for (const [k, v] of this.tokenCache) {
              if ((v.claims as any).sub === user_id) {
                this.tokenCache.delete(k);
              }
            }
          }
        } catch (err) { console.error('[keycloak] token invalidation parse error:', err instanceof Error ? err.message : err); }
      });
    } catch (err) { console.error('[keycloak] Redis subscription unavailable:', err instanceof Error ? err.message : err); }
  }

  async ping(): Promise<void> {
    const resp = await fetch(`${this.realmUrl}/.well-known/openid-configuration`);
    if (!resp.ok) throw new Error(`Keycloak unhealthy: ${resp.status}`);
  }

  async validateToken(token: string): Promise<Record<string, unknown>> {
    const cached = this.tokenCache.get(token);
    if (cached && cached.expiresAt > Date.now()) return cached.claims;

    const resp = await fetch(`${this.realmUrl}/protocol/openid-connect/userinfo`, { headers: { Authorization: `Bearer ${token}` } });
    if (!resp.ok) throw new Error(`Token invalid (${resp.status})`);
    const claims = await resp.json() as Record<string, unknown>;
    this.tokenCache.set(token, { claims, expiresAt: Date.now() + this.cacheTTL });

    // Evict expired entries (limit scan to avoid perf issues)
    let evicted = 0;
    for (const [k, v] of this.tokenCache) {
      if (v.expiresAt <= Date.now()) { this.tokenCache.delete(k); evicted++; }
      if (evicted >= 50) break;
    }
    return claims;
  }

  /** Invalidate a specific token across all service replicas via Redis pub/sub. */
  async invalidateToken(token: string): Promise<void> {
    this.tokenCache.delete(token);
    if (this.redisClient) {
      await this.redisClient.publish('__token_invalidation__', { token, timestamp: Date.now() });
    }
  }

  /** Invalidate all tokens for a user across all replicas (e.g. on password change or revocation). */
  async invalidateUserTokens(userId: string): Promise<void> {
    for (const [k, v] of this.tokenCache) {
      if ((v.claims as any).sub === userId) {
        this.tokenCache.delete(k);
      }
    }
    if (this.redisClient) {
      await this.redisClient.publish('__token_invalidation__', { user_id: userId, timestamp: Date.now() });
    }
  }

  getKYCLevel(claims: Record<string, unknown>): number {
    const attrs = claims.attributes as Record<string, unknown> | undefined;
    if (attrs?.kyc_level !== undefined) return Number(attrs.kyc_level);
    if (claims.kyc_level !== undefined) return Number(claims.kyc_level);
    return 0;
  }

  async getServiceToken(): Promise<string> {
    const body = new URLSearchParams({ grant_type: 'client_credentials', client_id: this.clientId, client_secret: this.clientSecret });
    const resp = await fetch(`${this.realmUrl}/protocol/openid-connect/token`, { method: 'POST', body, headers: { 'Content-Type': 'application/x-www-form-urlencoded' } });
    if (!resp.ok) throw new Error(`Service token failed (${resp.status})`);
    const data = await resp.json() as Record<string, unknown>;
    return data.access_token as string;
  }

  async updateUserKYCLevel(userId: string, kycLevel: number): Promise<void> {
    const token = await this.getServiceToken();
    const realm = this.realmUrl.split('/realms/')[1] || 'insurance';
    await fetch(`${this.adminUrl}/admin/realms/${realm}/users/${userId}`, {
      method: 'PUT', headers: { Authorization: `Bearer ${token}`, 'Content-Type': 'application/json' },
      body: JSON.stringify({ attributes: { kyc_level: [String(kycLevel)] } }),
    });
    // Invalidate cached tokens for this user across all replicas
    await this.invalidateUserTokens(userId);
  }
}

File: infrastructure/go-sdk/keycloak.go
package infra

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"go.uber.org/zap"
)

const maxTokenCacheSize = 10000

type KeycloakClient struct {
	realmURL     string
	clientID     string
	clientSecret string
	adminURL     string
	httpClient   *http.Client
	logger       *zap.Logger
	tokenCache   map[string]*cachedToken
	mu           sync.RWMutex
	redis        *RedisClient
}

type cachedToken struct {
	Claims    map[string]interface{}
	ExpiresAt time.Time
}

type KeycloakUser struct {
	ID         string                 `json:"id"`
	Username   string                 `json:"username"`
	Email      string                 `json:"email"`
	FirstName  string                 `json:"firstName"`
	LastName   string                 `json:"lastName"`
	Enabled    bool                   `json:"enabled"`
	Attributes map[string][]string    `json:"attributes,omitempty"`
}

func NewKeycloakClient(logger *zap.Logger, realmURL, clientID, clientSecret, adminURL string) *KeycloakClient {
	return &KeycloakClient{
		realmURL:     realmURL,
		clientID:     clientID,
		clientSecret: clientSecret,
		adminURL:     adminURL,
		httpClient:   &http.Client{Timeout: 10 * time.Second},
		logger:       logger,
		tokenCache:   make(map[string]*cachedToken),
	}
}

func (c *KeycloakClient) Ping(ctx context.Context) error {
	req, err := http.NewRequestWithContext(ctx, "GET", c.realmURL+"/.well-known/openid-configuration", nil)
	if err != nil {
		return err
	}
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("keycloak ping: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return fmt.Errorf("keycloak unhealthy: %d", resp.StatusCode)
	}
	return nil
}

func (c *KeycloakClient) ValidateToken(ctx context.Context, token string) (map[string]interface{}, error) {
	c.mu.RLock()
	if cached, ok := c.tokenCache[token]; ok && time.Now().Before(cached.ExpiresAt) {
		c.mu.RUnlock()
		return cached.Claims, nil
	}
	c.mu.RUnlock()

	req, err := http.NewRequestWithContext(ctx, "GET", c.realmURL+"/protocol/openid-connect/userinfo", nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("token validation: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("invalid token: status %d", resp.StatusCode)
	}
	body, _ := io.ReadAll(resp.Body)
	var claims map[string]interface{}
	json.Unmarshal(body, &claims)

	c.mu.Lock()
	c.tokenCache[token] = &cachedToken{Claims: claims, ExpiresAt: time.Now().Add(5 * time.Minute)}
	// Enforce max cache size — evict expired entries
	if len(c.tokenCache) > maxTokenCacheSize {
		now := time.Now()
		for k, v := range c.tokenCache {
			if now.After(v.ExpiresAt) {
				delete(c.tokenCache, k)
			}
		}
	}
	c.mu.Unlock()

	return claims, nil
}

func (c *KeycloakClient) GetKYCLevel(ctx context.Context, token string) (int, error) {
	claims, err := c.ValidateToken(ctx, token)
	if err != nil {
		return 0, err
	}
	if level, ok := claims["kyc_level"]; ok {
		switch v := level.(type) {
		case float64:
			return int(v), nil
		case string:
			var l int
			fmt.Sscanf(v, "%d", &l)
			return l, nil
		}
	}
	return 0, nil
}

func (c *KeycloakClient) GetServiceToken(ctx context.Context) (string, error) {
	data := url.Values{}
	data.Set("grant_type", "client_credentials")
	data.Set("client_id", c.clientID)
	data.Set("client_secret", c.clientSecret)

	req, err := http.NewRequestWithContext(ctx, "POST", c.realmURL+"/protocol/openid-connect/token", strings.NewReader(data.Encode()))
	if err != nil {
		return "", err
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return "", fmt.Errorf("service token: %w", err)
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	var result map[string]interface{}
	json.Unmarshal(body, &result)
	if token, ok := result["access_token"].(string); ok {
		return token, nil
	}
	return "", fmt.Errorf("no access_token in response")
}

func (c *KeycloakClient) UpdateUserKYCLevel(ctx context.Context, userID string, kycLevel int, kycStatus string) error {
	adminToken, err := c.GetServiceToken(ctx)
	if err != nil {
		return err
	}
	payload := map[string]interface{}{
		"attributes": map[string][]string{
			"kyc_level":  {fmt.Sprintf("%d", kycLevel)},
			"kyc_status": {kycStatus},
		},
	}
	data, _ := json.Marshal(payload)
	realm := extractRealm(c.realmURL)
	req, err := http.NewRequestWithContext(ctx, "PUT", fmt.Sprintf("%s/admin/realms/%s/users/%s", c.adminURL, realm, userID), bytes.NewReader(data))
	if err != nil {
		return err
	}
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", "Bearer "+adminToken)
	resp, err := c.httpClient.Do(req)
	if err != nil {
		return fmt.Errorf("update user: %w", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode >= 400 {
		body, _ := io.ReadAll(resp.Body)
		return fmt.Errorf("update user failed (%d): %s", resp.StatusCode, string(body))
	}
	return nil
}

// SetupRealmClients registers all platform service clients in the Keycloak realm.
func (c *KeycloakClient) SetupRealmClients(ctx context.Context) error {
	clients := []map[string]interface{}{
		{"clientId": "policy-service", "name": "Policy Service", "serviceAccountsEnabled": true, "directAccessGrantsEnabled": false},
		{"clientId": "claims-service", "name": "Claims Service", "serviceAccountsEnabled": true, "directAccessGrantsEnabled": false},
		{"clientId": "payment-service", "name": "Payment Service", "serviceAccountsEnabled": true, "directAccessGrantsEnabled": false},
		{"clientId": "kyc-service", "name": "KYC Orchestrator", "serviceAccountsEnabled": true, "directAccessGrantsEnabled": false},
		{"clientId": "fraud-service", "name": "Fraud Detection", "serviceAccountsEnabled": true, "directAccessGrantsEnabled": false},
		{"clientId": "analytics-service", "name": "Analytics Service", "serviceAccountsEnabled": true, "directAccessGrantsEnabled": false},
		{"clientId": "mobile-app", "name": "Mobile Application", "publicClient": true, "redirectUris": []string{"*"}},
		{"clientId": "customer-portal", "name": "Customer Portal", "publicClient": true, "redirectUris": []string{"*"}},
		{"clientId": "agent-portal", "name": "Agent Portal", "publicClient": true, "redirectUris": []string{"*"}},
	}

	adminToken, err := c.GetServiceToken(ctx)
	if err != nil {
		return err
	}
	realm := extractRealm(c.realmURL)
	for _, client := range clients {
		data, _ := json.Marshal(client)
		req, _ := http.NewRequestWithContext(ctx, "POST", fmt.Sprintf("%s/admin/realms/%s/clients", c.adminURL, realm), bytes.NewReader(data))
		req.Header.Set("Content-Type", "application/json")
		req.Header.Set("Authorization", "Bearer "+adminToken)
		resp, err := c.httpClient.Do(req)
		if err != nil {
			c.logger.Warn("client_registration_failed", zap.String("client", client["clientId"].(string)), zap.Error(err))
			continue
		}
		resp.Body.Close()
	}
	return nil
}

// SetRedisClient attaches Redis for distributed token invalidation across replicas.
func (c *KeycloakClient) SetRedisClient(redis *RedisClient) {
	c.redis = redis
	go c.subscribeToInvalidations()
}

func (c *KeycloakClient) subscribeToInvalidations() {
	if c.redis == nil || c.redis.client == nil {
		return
	}
	ctx := context.Background()
	sub := c.redis.Subscribe(ctx, "__token_invalidation__")
	defer sub.Close()

	for msg := range sub.Channel() {
		var data map[string]interface{}
		if err := json.Unmarshal([]byte(msg.Payload), &data); err != nil {
			continue
		}
		if token, ok := data["token"].(string); ok && token != "" {
			c.mu.Lock()
			delete(c.tokenCache, token)
			c.mu.Unlock()
		}
		if userID, ok := data["user_id"].(string); ok && userID != "" {
			c.mu.Lock()
			for k, v := range c.tokenCache {
				if sub, ok := v.Claims["sub"].(string); ok && sub == userID {
					delete(c.tokenCache, k)
				}
			}
			c.mu.Unlock()
		}
	}
}

func (c *KeycloakClient) InvalidateTokenCache(token string) {
	c.mu.Lock()
	delete(c.tokenCache, token)
	c.mu.Unlock()
}

// InvalidateToken invalidates a specific token across all service replicas via Redis pub/sub.
func (c *KeycloakClient) InvalidateToken(ctx context.Context, token string) error {
	c.InvalidateTokenCache(token)
	if c.redis != nil {
		event := map[string]interface{}{"token": token, "timestamp": time.Now().Unix()}
		return c.redis.Publish(ctx, "__token_invalidation__", event)
	}
	return nil
}

// InvalidateUserTokens invalidates all tokens for a user across all replicas.
func (c *KeycloakClient) InvalidateUserTokens(ctx context.Context, userID string) error {
	c.mu.Lock()
	for k, v := range c.tokenCache {
		if sub, ok := v.Claims["sub"].(string); ok && sub == userID {
			delete(c.tokenCache, k)
		}
	}
	c.mu.Unlock()
	if c.redis != nil {
		event := map[string]interface{}{"user_id": userID, "timestamp": time.Now().Unix()}
		return c.redis.Publish(ctx, "__token_invalidation__", event)
	}
	return nil
}

func extractRealm(realmURL string) string {
	parts := strings.Split(realmURL, "/realms/")
	if len(parts) == 2 {
		return parts[1]
	}
	return "insurance"
}
